core/unit_test/TestMathematicalFunctions.hpp

```diff
@@ -1791,6 +1795,126 @@ struct TestIsNaN {
+  TestNextAfterHalf() { run(); }
+  void run() const {
+    int errors = 0;
+    Kokkos::parallel_reduce(Kokkos::RangePolicy<Space>(0, 1), *this, errors);
+    ASSERT_EQ(errors, 0);
+  }
+  KOKKOS_FUNCTION void operator()(int, int& e) const {
+    using KE::infinity;
+    using KE::quiet_NaN;
+    using KE::signaling_NaN;
+    using Kokkos::isnan;
+    using Kokkos::nextafter;
+
+    // Define useful constants
+    const std::uint16_t FP16_POS_ZERO     = 0x0000;
+    const std::uint16_t FP16_NEG_ZERO     = 0x8000;
+    const std::uint16_t FP16_SMALLEST_POS = 0x0001;
+    const std::uint16_t FP16_SMALLEST_NEG = 0x8001;
+
+    const FP16Type pos_one{1.0f}, pos_two{2.0f};
+    const FP16Type neg_one{-1.0f}, neg_two{-2.0f};
+    const FP16Type pos_zero     = Kokkos::bit_cast<FP16Type>(FP16_POS_ZERO);
+    const FP16Type neg_zero     = Kokkos::bit_cast<FP16Type>(FP16_NEG_ZERO);
+    const FP16Type pos_smallest = Kokkos::bit_cast<FP16Type>(FP16_SMALLEST_POS);
+    const FP16Type neg_smallest = Kokkos::bit_cast<FP16Type>(FP16_SMALLEST_NEG);
+    const FP16Type pos_max = Kokkos::Experimental::finite_max<FP16Type>::value;
+    const FP16Type neg_max = Kokkos::Experimental::finite_min<FP16Type>::value;
+    const FP16Type pos_inf = Kokkos::Experimental::infinity<FP16Type>::value;
+    const FP16Type neg_inf =
+        -static_cast<FP16Type>(Kokkos::Experimental::infinity<FP16Type>::value);
+
+    // NaN Handling
+    if (!isnan(nextafter(quiet_NaN<FP16Type>::value, pos_one)) ||
+        !isnan(nextafter(signaling_NaN<FP16Type>::value, pos_one)) ||
+        !isnan(nextafter(pos_one, quiet_NaN<FP16Type>::value)) ||
+        !isnan(nextafter(pos_one, signaling_NaN<FP16Type>::value)) ||
+        !isnan(nextafter(quiet_NaN<FP16Type>::value,
+                         quiet_NaN<FP16Type>::value)) ||
+        !isnan(nextafter(quiet_NaN<FP16Type>::value,
+                         signaling_NaN<FP16Type>::value)) ||
+        !isnan(nextafter(signaling_NaN<FP16Type>::value,
+                         quiet_NaN<FP16Type>::value)) ||
+        !isnan(nextafter(signaling_NaN<FP16Type>::value,
+                         signaling_NaN<FP16Type>::value))) {
+      ++e;
+      Kokkos::printf("failed half precision nextafter(NaN)\n");
+    }
+
+    // Equality (from==toward) Handling
+    if (nextafter(pos_one, pos_one) != pos_one ||
+        nextafter(pos_zero, pos_zero) != pos_zero ||
+        nextafter(neg_zero, neg_zero) != neg_zero ||
+        nextafter(pos_inf, pos_inf) != pos_inf ||
+        nextafter(neg_inf, neg_inf) != neg_inf) {
+      ++e;
+      Kokkos::printf("failed half precision nextafter(equality)\n");
+    }
+
+    // Zero Handling
+    if (nextafter(pos_zero, pos_one) != pos_smallest ||
+        nextafter(pos_zero, neg_one) != neg_smallest ||
+        nextafter(neg_zero, pos_one) != pos_smallest ||
+        nextafter(neg_zero, neg_one) != neg_smallest) {
+      ++e;
+      Kokkos::printf("failed half precision nextafter(zero)\n");
+    }
```

Comment:
Do we want to check that `nextafter(neg_zero, pos_zero)` returns `pos_zero` and that `nextafter(pos_zero, neg_zero)` returns `neg_zero` conversely?

Reply:
Good catch. I will add them